design/ingressroute-design.md

```diff
@@ -0,0 +1,300 @@
+    delegate:
+      # the name of the IngressRoute object to delegate to
+      name: google-finance
+      # the namespace of the IngressRoute object, if blank the current namespace is assumed
+      namespace: finance
+  - match: /ads
+    # more than one service name/port pair may be present allowing the services for a route to
+    # be handled by more than one service.
+    service:
+    - name: ads-red
+      port: 9090
+    - name: ads-blue
+      port: 9090
+```
+
+This is an example of the `google-finance` object which has been delegated responsibilty for paths starting with `/finance`.
+
+```yaml
+apiVersion: contour.heptio.com/v1alpha1
+kind: IngressRoute
+metadata:
+  name: google-finance
+  namespace: finance
+spec:
+  # note that this is an edge, so there is no virtualhost key
+  # routes contains the set of routes for this virtual host.
+  # routes must _always_ be present and non empty.
+  # routes can be present in any order, and will be matched from most to least
+  # specific, however as this is an edge, only prefixes that match the prefix
+  # that delegated to this object.
+  routes:
+  # each route in this object must start with /finance as this was the prefix which
+  # delegated to this object.
+  - match: /finance
+    service:
+    - name: finance-app
+      port: 9999
+  - match: /finance/static
+    service:
+    - name: finance-static-content
+      port: 8080
+  - match: /finance/partners
+    # delegate the /finance/partners prefix to the partners namespace
+    delegate:
+      name: finance-partners
+      namespace: partners
+```
+
+## Delegation rules
+
+The delegation rules applied are as follows
+
+1. If an `IngressRoute` object contains a `spec.virtualhost` key it is considered a root.
+2. If an `IngressRoute` object does not contain `spec.virtualhost` key is considered an edge.
+3. An edge is reachable if a delegation to it exists in another `IngressRoute` object.
+4. Edges which are not reachable are considered orphened. Orphened edges have no effect on the running configuration.
+
+## Validation rules
+
+The validation rules applied in this design are as follows
+Some of these rules may be relaxed in future designs.
+
+1. If a validation error is encountered, the entire object is rejected. Partial application of the valid portions of the configuration is _not_ attempted.
+2. The prefix of the delegate route in the parent must be match the routes in the child.
+  ```yaml
+apiVersion: contour.heptio.com/v1alpha1
+kind: IngressRoute
+metadata:
+  name: root
+  namespace: root
+spec:
+  virtualhost:
+    fqdn: heptio.com
+    aliases:
+      - www.heptio.com
+  routes:
+  - match: /static
+    delegate:
+      name: child
+      namespace: child
+---
+apiVersion: contour.heptio.com/v1alpha1
+kind: IngressRoute
+metadata:
+  name: child 
+  namespace: child
+spec:
+  routes:
+  # valid
+  - match: /static/css
+  # invalid
+  - match: /css
```

Comment:
Wouldn't `/css` just be an orphan since no root delegates to it (yet)? The issue is you break all the routes defined and could cause traffic problems. Thinking it would be nice to mark this route as orphaned and traffic won't route anyway since it's not delegated properly.

Reply:
I think having duplicate `matches` should surface an error since there's no happy path forward for the routes.

Reply:
> Wouldn't /css just be an orphan since no root delegates to it (yet)? 

It could be, but as a simplification unless all the routes mentioned match the prefix you're delegated on, the entire object is invalid. This rule may be relaxed in the future, although reporting status on a per spec.route basis is going to make for a pretty hard to read delegationStatus.

Reply:
If we make this stuff relative then we won't have these issues.

Eliminating all routes based on one route going invalid will be non-intuitive and probably bad.  Let's say the parent was redirecting both `/static` and `/css` and then removes the delegation for `/css`.  That will also break `/static`.  My gut is that any linkages should be looser than that.

So -- in this case the *object* is valid but the graph is somewhat confusing and invalid.  I think partial application is the right thing in that case.

Reply:
@jbeda I don't want to entertain relative routes at this stage. We could add them later, but there are lots of edge cases which I don't want to get hung up on for this design round. I don't think we loose anything by starting strict and opening up as we go on, but the inverse is less true.

In the same vein, by treating configuration on a _per object_, not _per route_ basis, this makes validation simpler for this round of design. We could move from a per object to per route model in the future, but the opposite is harder. The former, the object applies or it doesn't, the latter an object which was "sort of " valid suddenly gets rejected outright. I don't want to cut off that flexibility too early.

Partial application is straight forward to do, but reporting this status in a realistic way -- not a giant-ass `status:` blob -- is harder to do.